src/node_watchdog.cc

```diff
@@ -100,16 +124,55 @@ SigintWatchdog::SigintWatchdog(
   SigintWatchdogHelper::GetInstance()->Start();
 }
 
+SigintWatchdog::SigintWatchdog(
+  v8::Isolate* isolate, Local<v8::Function> interrupt)
+    : isolate_(isolate) {
+  ref.Reset(isolate, interrupt);
+  auto env = node::Environment::GetCurrent(isolate);
+  loop_activate = new uv_async_t;
+  loop_activate->data = this;
+  // Just activate the loop, doesn't do callback.
+  uv_async_init(env->event_loop(), loop_activate, [](uv_async_t* it){
+    SigintWatchdog* self = reinterpret_cast<SigintWatchdog*>(it->data);
+    self->HandleInterrupt();
+  });
+  uv_unref(reinterpret_cast<uv_handle_t*>(loop_activate));
+  // Register this watchdog with the global SIGINT/Ctrl+C listener.
+  SigintWatchdogHelper::GetInstance()->Register(this);
+  // Start the helper thread, if that has not already happened.
+  SigintWatchdogHelper::GetInstance()->Start();
+}
 
 SigintWatchdog::~SigintWatchdog() {
+  if (loop_activate != nullptr) {
+    uv_close(reinterpret_cast<uv_handle_t*>(loop_activate),
+      SigintWatchdog::CloseAsync);
+  }
+  ref.Reset();
   SigintWatchdogHelper::GetInstance()->Unregister(this);
   SigintWatchdogHelper::GetInstance()->Stop();
 }
 
-
 void SigintWatchdog::HandleSigint() {
-  *received_signal_ = true;
-  isolate_->TerminateExecution();
+  if (ref.IsEmpty()) {
+    *received_signal_ = true;
+    isolate_->TerminateExecution();
+  } else {
+    uv_async_send(loop_activate);
+    isolate_->RequestInterrupt([](v8::Isolate* isolate, void* data) {
+      SigintWatchdog* self = reinterpret_cast<SigintWatchdog*>(data);
+      self->HandleInterrupt();
+    }, this);
+  }
+}
+
+void SigintWatchdog::HandleInterrupt() {
+  v8::HandleScope handle_scope(isolate_);
+  auto callback = ref.Get(isolate_);
+  CHECK_EQ(callback.IsEmpty(), false);
+  CHECK_EQ(callback->IsFunction(), true);
+  auto recv = v8::Object::New(isolate_);
+  callback->Call(callback->CreationContext(), recv, 0, {});
```

Comment:
You don’t technically need a receiver here.

If you do take the above suggestion and switch to using `HandleWrap()`, the typical pattern would be to call a method on the object rather than passing one to the constructor, e.g.

```c++
MakeCallback(env->oninterrupt_string(), 0, nullptr);
```

and in JS

```js
const watchdog = new SigintWatchdog();
watchdog.oninterrupt = () => {
  ...
};
```

(i.e. in that case you don’t need to manage `ref` explicitly.)